org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MavenPluginActivator.java

```diff
@@ -216,14 +219,13 @@ public void start(final BundleContext context) throws Exception {
 
     this.mavenBackgroundJob = new ProjectRegistryRefreshJob(managerImpl, mavenConfiguration);
 
-    IWorkspace workspace = ResourcesPlugin.getWorkspace();
-    workspace.addResourceChangeListener(mavenBackgroundJob,
-        IResourceChangeEvent.POST_CHANGE | IResourceChangeEvent.PRE_CLOSE | IResourceChangeEvent.PRE_DELETE);
+    context.registerService(IResourceChangeListener.class, mavenBackgroundJob, getMaskProperties(
+        IResourceChangeEvent.POST_CHANGE | IResourceChangeEvent.PRE_CLOSE | IResourceChangeEvent.PRE_DELETE));
 
     this.projectManager = new MavenProjectManager(managerImpl, mavenBackgroundJob, stateLocationDir);
     this.projectManager.addMavenProjectChangedListener(new WorkspaceStateWriter(projectManager));
     if(updateProjectsOnStartup || managerImpl.getProjects().length == 0) {
-      this.projectManager.refresh(new MavenUpdateRequest(workspace.getRoot().getProjects(), //
+      this.projectManager.refresh(new MavenUpdateRequest(ResourcesPlugin.getWorkspace().getRoot().getProjects(), //
```

Comment:
Can't we get it from context then?

Reply:
Yeah, I just aimed to fix the obvious ones first.